File: _CodeTopics/LeetCode_contest/weekly/weekly2020/205/205_1.py
class Solution(object):
    def modifyString(self, s):
        """
        :type s: str
        :rtype: str
        """
        
        s = list(s)
        length = len(s)
        if s[0] == '?':
            s[0] = 'a'
        for i in range(1, length):
            if s[i] == '?':
                if s[i-1] != 'a':
                    s[i] = 'a'
                else:
                    s[i] = 'b'
            else:
                if s[i] == s[i-1]:
                    # 此时s[i-1]只能是从问号变过来的。
                    s[i-1] = 'c'
        return ''.join(s)
    
"""
https://leetcode-cn.com/submissions/detail/105224224/

100 / 100 个通过测试用例
状态：通过
执行用时: 24 ms
内存消耗: 12.7 MB
"""

File: _CodeTopics/LeetCode/1401-1600/001576--weekly205_1/tran/205_1.cpp
class Solution {
public:
    string modifyString(string s) {
        
        // 第 205 场周赛第一题
        // https://leetcode-cn.com/submissions/detail/105224224/

        int length = s.size();
        if (s[0] == '?')
            s[0] = 'a';
        for (int i = 1; i < length; ++i) {
            if (s[i] == '?') {
                if (s[i-1] != 'a')
                    s[i] = 'a';
                else
                    s[i] = 'b';
            } else {
                if (s[i] == s[i-1]) {
                    if (i > 1) {
                        if (s[i-2] != 'a' && s[i] != 'a')
                            s[i-1] = 'a';
                        else if (s[i-2] != 'b' && s[i] != 'b')
                            s[i-1] = 'b';
                        else
                            s[i-1] = 'c';
                    } else {
                        s[i-1] = 'b';
                    }
                }
            }
        }
        return s;
    }
};

/*
https://leetcode-cn.com/submissions/detail/255050207/

执行用时：4 ms, 在所有 C++ 提交中击败了43.75%的用户
内存消耗：5.9 MB, 在所有 C++ 提交中击败了67.97%的用户
通过测试用例：
776 / 776
*/
